### PR TITLE
AMQP-486 Convert from DocBook to AsciiDoc

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-486

Initial Conversion using https://github.com/garyrussell/docbook2asciidoctor/tree/amqp with minor manual tweaks.

Ready for initial review.

    ./gradlew reference

### CODE REVIEW

**Comment by artembilan** [comment]
@garyrussell , how to review this, if all those XMLs are still here ?
Just take a look to that new `epub` ?

**Comment by artembilan** [comment]
To be honest, looks ugly :smile: 
amqp-epub

**Comment by garyrussell** [comment]
I didn't look at the ePub book yet; I think the first step to review is to review side-by-side the generated docs (PDF and HTML).

File: build.gradle [hunk 1]
```diff
@@ -9,12 +9,15 @@ buildscript {
 		maven { url 'http://repo.springsource.org/plugins-release' }
 	}
 	dependencies {
-		classpath 'org.springframework.build.gradle:docbook-reference-plugin:0.2.8'
+		classpath 'io.spring.gradle:docbook-reference-plugin:0.3.0'
+		classpath 'org.asciidoctor:asciidoctor-gradle-plugin:1.5.0'
```
Comment (artembilan) [inline build.gradle]: Aren't we going to go ahead with the newer `asciidoctor` plugin version: https://github.com/rwinch/asciidoctor-gradle-examples/blob/master/asciidoc-to-all-example/build.gradle?

File: src/reference/asciidoc/preface.adoc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[[preface]]
+The Spring AMQP project applies core Spring concepts to the development of AMQP-based messaging solutions.
+We provide a "template" as a high-level abstraction for sending and receiving messages.
+We also provide support for Message-driven POJOs.
+These libraries facilitate management of AMQP resources while promoting the use of dependency injection and declarative configuration.
+In all of these cases, you will see similarities to the JMS support in the Spring Framework.
+The project consists of both Java and .NET versions.
+This manual is dedicated to the Java version.
+For links to the .NET version's manual or any other project-related information visit the Spring AMQP project http://www.springsource.org/spring-amqp[homepage].
```
Comment (artembilan) [inline src/reference/asciidoc/preface.adoc]: `http://projects.spring.io/spring-amqp/`

**Comment by artembilan** [comment]
`amqp.adoc` isn't visible here to comment, hence I do that here in the common comment:
1. line 365 `Routing Connection Factory` - `[source,xml]` must be on the new line closer to the `----`. And here for the next code block `[source,java]` without indents
2. Line 1640 `Declarative queues must have` - `[source,xml]` from the new line.
3. `Configuration options for a message listener container` table needs borders.

**Comment by artembilan** [comment]
That's all for HTML. I removed `docbook` dir locally and it works well.

File: src/reference/asciidoc/whats-new.adoc [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+===== Listener Queues
+
+The listener container now, by default, redeclares any missing queues during startup.
+A new `auto-declare` attribute has been added to the `<rabbit:listener-container>` to prevent these redeclarations.
+See <<lc-auto-delete>>.
+
+===== RabbitTemplate: mandatory and connectionFactorySelector Expressions
+
+The `mandatoryExpression` and `sendConnectionFactorySelectorExpression` and `receiveConnectionFactorySelectorExpression` SpEL `Expression`s properties have been added to the `RabbitTemplate`.
```
Comment (artembilan) [inline src/reference/asciidoc/whats-new.adoc]: Need to fix to this "`Expressions`" I guess.

File: src/reference/asciidoc/whats-new.adoc [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+Also, the container will now start if at least one of its configured queues is available for use.
+See <<listener-queues>>
+
+This listener container will now redeclare any auto-delete queues during startup.
+See <<lc-auto-delete>>.
+
+===== Consumer Priority
+
+The listener container now supports consumer arguments, allowing the `x-priority` argument to be set.
+See <<consumer-priority>>.
```
Comment (artembilan) [inline src/reference/asciidoc/whats-new.adoc]: There is no such a `consumer-priority` anchor.

File: src/reference/asciidoc/whats-new.adoc [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+===== Caching Connection Factory
+
+The caching connection factory can now be configured to cache `Connection`s and their `Channel`s instead of using a single connection and caching just `Channel`s.
+See <<connections>>.
+
+===== Binding Arguments
+
+The `<exchange>`'s `<binding>` now supports parsing of the `<binding-arguments>` sub-element.
+The `<headers-exchange>`'s `<binding>` now can be configured with a `key/value` attribute pair (to match on a single header) or with a `<binding-arguments>` sub-element, allowing matching on multiple headers; these options are mutually exclusive.
+See <<headers-exchange>>.
```
Comment (artembilan) [inline src/reference/asciidoc/whats-new.adoc]: There is no such a `headers-exchange` anchor

**Comment by artembilan** [comment]
The similar issue around a `[source,java]` for the `receiveAndReply` code block - `processing and replying to messages`

**Comment by artembilan** [comment]
There is a code style issue for the phrase:
> With the default MessageConverter`s, the method paramters

**Comment by artembilan** [comment]
And:
> then all RabbitAdmin`s will declare the object (as long as

**Comment by artembilan** [comment]
`Configuration options for a message listener container` table looks good for me in the PDF, although I don't know if there should be a zebra

**Comment by artembilan** [comment]
That's all.
Let me know for what I should pay attention

File: build.gradle [hunk 6]
```diff
@@ -233,12 +236,40 @@ project('spring-rabbit') {
...
+		]
+	]
+}
+
+apply plugin: DocbookReferencePlugin
 
 reference {
-	sourceDir = file('src/reference/docbook')
+	sourceFileName = 'index.xml'
+	sourceDir = file("$buildDir/html")
```
Comment (artembilan) [inline build.gradle]: Redundant. Can be like:
```
sourceDir = asciidoctor.outputDir
```

File: build.gradle [hunk 7]
```diff
@@ -233,12 +236,40 @@ project('spring-rabbit') {
 
 }
 
-apply plugin: 'docbook-reference'
+apply plugin: org.asciidoctor.gradle.AsciidoctorPlugin
+
+asciidoctor {
+	sourceDir file("$docsDir")
+	sourceDocumentNames = files("$docsDir/index.adoc") // Change in >= 1.5.1
+	outputDir file("$buildDir/html")
```
Comment (artembilan) [inline build.gradle]: Redundant. We can just rely on the default one.

File: build.gradle [hunk 8]
```diff
@@ -233,12 +236,40 @@ project('spring-rabbit') {
...
+			imagesdir: '',
+			stylesdir: "stylesheets/",
+			stylesheet: 'golo.css',
+			'spring-amqp-version': "$version",
+			'source-highlighter': 'highlightjs'
+		]
+	]
+}
+
+apply plugin: DocbookReferencePlugin
```
Comment (artembilan) [inline build.gradle]: This:
```
apply plugin: "docbook-reference"
```
for me works well.

File: build.gradle (hunk 1 above)
Comment (garyrussell) [inline build.gradle]: Cool - I pulled all this code from XD

File: src/reference/asciidoc/preface.adoc (hunk 2 above)
Comment (garyrussell) [inline src/reference/asciidoc/preface.adoc]: Actually, the spring.io page has no links to .NET versions.

File: src/reference/asciidoc/whats-new.adoc (hunk 3 above)
Comment (garyrussell) [inline src/reference/asciidoc/whats-new.adoc]: Or `Expression` s (extra space)

File: src/reference/asciidoc/preface.adoc (hunk 2 above)
Comment (artembilan) [inline src/reference/asciidoc/preface.adoc]: But `http://www.springsource.org/spring-amqp` redirects to that IO page...

Comment (garyrussell) [inline src/reference/asciidoc/preface.adoc]: Yes; it should be changed; I am just saying we should drop the discussion about .NET

**Comment by garyrussell** [comment]
FYI, I fixed the conversion script (for SI) to put `[source]` on a new line.

**Comment by garyrussell** [comment]
Pushed; I don't yet have a solution for the table cell borders.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/08e88d465ff1dcd655cbdb3ff30022a930c90ae5 after rebase and polishing. 
See the commit message.

**Pull request closed without merging**